race condition when stopping muxer

```
panic: close of closed channel

goroutine 1078 [running]:
github.com/cloudstruct/go-ouroboros-network/muxer.(*Muxer).Stop(0x4000076e80)
        /go/pkg/mod/github.com/cloudstruct/go-ouroboros-network@v0.13.3/muxer/muxer.go:57 +0x94
github.com/cloudstruct/go-ouroboros-network.(*Ouroboros).Close(0x4000324460)
        /go/pkg/mod/github.com/cloudstruct/go-ouroboros-network@v0.13.3/ouroboros.go:107 +0x30
github.com/cloudstruct/go-ouroboros-network.(*Ouroboros).setupConnection.func2(0x4000324460)
        /go/pkg/mod/github.com/cloudstruct/go-ouroboros-network@v0.13.3/ouroboros.go:175 +0xb4
created by github.com/cloudstruct/go-ouroboros-network.(*Ouroboros).setupConnection
        /go/pkg/mod/github.com/cloudstruct/go-ouroboros-network@v0.13.3/ouroboros.go:171 +0x254
```

--- a/muxer/muxer.go
+++ b/muxer/muxer.go
@@ -22,6 +22,7 @@ type Muxer struct {
 	sendMutex         sync.Mutex
 	startChan         chan bool
 	doneChan          chan bool
+	stopMutex         sync.Mutex
 	ErrorChan         chan error
 	protocolSenders   map[uint16]chan *Segment
 	protocolReceivers map[uint16]chan *Segment
@@ -45,12 +46,18 @@ func (m *Muxer) Start() {
 }
 
 func (m *Muxer) Stop() {
+	// We use a mutex to prevent this function from being called multiple times
+	// concurrently, which would cause a race condition
+	m.stopMutex.Lock()
+	defer m.stopMutex.Unlock()
 	// Immediately return if we're already shutting down
 	select {
 	case <-m.doneChan:
 		return
 	default:
 	}
+	// Close doneChan to signify that we're shutting down
+	close(m.doneChan)
 	// Close protocol receive channels
 	// We rely on the individual mini-protocols to close the sender channel
 	for _, recvChan := range m.protocolReceivers {
@@ -58,8 +65,6 @@ func (m *Muxer) Stop() {
 	}
 	// Close ErrorChan to signify to consumer that we're shutting down
 	close(m.ErrorChan)
-	// Close doneChan to signify that we're shutting down
-	close(m.doneChan)
 }
 
 func (m *Muxer) sendError(err error) {

